pkg/config/validation.go

```diff
@@ -0,0 +1,347 @@
+	}
+
+	// parse tolerations
+	for _, t := range config.Sync.ToHost.Pods.EnforceTolerations {
+		_, err := toleration.ParseToleration(t)
+		if err != nil {
+			return err
+		}
+	}
+
+	// check if enable scheduler works correctly
+	if config.ControlPlane.Advanced.VirtualScheduler.Enabled && !config.Sync.FromHost.Nodes.SyncAll && len(config.Sync.FromHost.Nodes.Selector.Labels) == 0 {
+		config.Sync.FromHost.Nodes.SyncAll = true
+	}
+
+	// enable additional controllers required for scheduling with storage
+	if config.ControlPlane.Advanced.VirtualScheduler.Enabled && config.Sync.ToHost.PersistentVolumeClaims.Enabled {
+		config.Sync.FromHost.CSINodes.Enabled = true
+		config.Sync.FromHost.CSIStorageCapacities.Enabled = true
+		config.Sync.FromHost.CSIDrivers.Enabled = true
+		if !config.Sync.FromHost.StorageClasses.Enabled && !config.Sync.ToHost.StorageClasses.Enabled {
+			config.Sync.FromHost.StorageClasses.Enabled = true
+		}
+	}
+
+	// check if nodes controller needs to be enabled
+	if config.ControlPlane.Advanced.VirtualScheduler.Enabled && !config.Sync.FromHost.Nodes.Enabled {
+		return fmt.Errorf("sync.fromHost.nodes.enabled is false, but required if using virtual scheduler")
+	}
+
+	// check if storage classes and host storage classes are enabled at the same time
+	if config.Sync.FromHost.StorageClasses.Enabled && config.Sync.ToHost.StorageClasses.Enabled {
+		return fmt.Errorf("you cannot enable both sync.fromHost.storageClasses.enabled and sync.toHost.storageClasses.enabled at the same time. Choose only one of them")
+	}
+
+	// validate central admission control
+	err := validateCentralAdmissionControl(config)
+	if err != nil {
+		return err
+	}
+
+	// validate generic sync config
+	err = validateGenericSyncConfig(config.Experimental.GenericSync)
+	if err != nil {
+		return fmt.Errorf("validate experimental.genericSync")
+	}
+
+	// validate distro
+	err = validateDistro(config)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateDistro(config *VirtualClusterConfig) error {
+	enabledDistros := 0
+	if config.Config.ControlPlane.Distro.K3S.Enabled {
+		enabledDistros++
+	}
+	if config.Config.ControlPlane.Distro.K0S.Enabled {
+		enabledDistros++
+	}
+	if config.Config.ControlPlane.Distro.K8S.Enabled {
+		enabledDistros++
+	}
+	if config.Config.ControlPlane.Distro.EKS.Enabled {
+		enabledDistros++
+	}
+
+	if enabledDistros > 1 {
+		return fmt.Errorf("please only enable a single distribution")
+	}
+	return nil
+}
+
+func validateGenericSyncConfig(config config.ExperimentalGenericSync) error {
+	err := validateExportDuplicates(config.Exports)
+	if err != nil {
+		return err
+	}
+
+	for idx, exp := range config.Exports {
+		if exp == nil {
+			return fmt.Errorf("exports[%d] is required", idx)
+		}
+
+		if exp.Kind == "" {
+			return fmt.Errorf("exports[%d].kind is required", idx)
+		}
+
+		if exp.APIVersion == "" {
+			return fmt.Errorf("exports[%d].APIVersion is required", idx)
+		}
```

Comment:
can't we have that from the json schema directly?

Reply:
Better to do valdiation here as well, also I haven't changed this